docs/thread_design_rfc.md

```diff
@@ -0,0 +1,69 @@
+## Current state
+
+NOTE: I used current names of threads, which are quite chaotic.
+It would be nice to use one uniform style.
+
+#### pageserver
+Main entry point for the pageserver executable.
+At start pageserver sets up logging, auth, metrics collection and starts following threads:
+
+- http_endpoint_thread
+- Page Service thread
+
+#### http_endpoint_thread
+Thread that serves incoming service requests, such as create tenant/branch, list tenants/branches.
+
+#### Page Service thread
+Listens for connections, and launches a new `conn_handler` handler thread for each.
+
+#### conn_handler() // these threads are not named
+Per connection handler.
+Communicates with one PostgresBackend using libpq protocol. Serves queries.
+
+#### Threads that run per tenant (repository)
+-------------
+
+#### wal_redo_manager
+Runs a child process with postgres binary in wal-redo mode.
+
+#### WAL receiver thread
+Connects to the WAL service, streams, parses and saves WAL in the repository.
+Starts on callmemaybe request.
+
+#### Checkpointer
+Flushes in-memory layers to disk.
+Currently it is launched at pageserver start only for pre-existing tenants.
+
+#### GC
+Removes on-disk layers that are not visible anymore.
+Currently it is launched at pageserver start only for pre-existing tenants.
+
+#### relish_uploader
```

Comment:
Just a note: current code placement for this kind of threads was selected naively.
A follow-up set of changes in the [branch](https://github.com/zenithdb/zenith/compare/kb/download_storage?expand=1) runs `relish_uploader` threads the same way as `http_endpoint_thread` is run.